PicPi/FrameAPIImpl.swift

```diff
@@ -0,0 +1,118 @@
+//
+//  FrameAPIImpl.swift
+//  PicPi
+//
+//  Created by Mustafa T. Mohammed on 12/25/20.
+//
+
+import Foundation
+import Moya
+import Promises
+
+class FrameAPIImpl: FrameAPI {
+  
+  func retrievePIPrefrences() -> Promise<[String : Any]> {
+    let promise = Promise<[String : Any]>.pending()
+    /// get the current PI preferencses as a dictionary
+    self.photoProvider.request(.retrievePIPreferences) { (result) in
+      switch result {
+        case .success(let response):
+          do {
+            /// trying to Serialize the response
+            let responseSerialized = try JSONSerialization.jsonObject(
+              with: response.data,
+              options: .mutableLeaves)
+            /// casting the serialized response (responseSerialized) as a dictionary
+            if let prefrencesDict = responseSerialized as? Dictionary<String,Any> {
```

Comment:
If this condition is false the promise is neither rejected nor resolved

You'll need to introduce a custom error for this case like so:

```
enum FrameAPIErrors: Error {
  case dataMalformed
}
```

Reply:
Great!